### PR TITLE
Added .net clients

### LINKED ISSUE #671: Add documentation for the .NET  client[FEATURE]
### Is your feature request related to a problem?
Add documentation for the .NET  client

https://www.nuget.org/profiles/opensearchproject

### PR DESCRIPTION
**Pull request by kolchfa-aws**
Fixes #671 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by joshuali925** [comment]
@MaxKsyunz @Yury-Fridlyand when you get a chance could you take a look at docs for .net clients? thanks

File: _clients/OSC-dot-net.md [hunk 1]
```diff
@@ -0,0 +1,330 @@
+---
+layout: default
+title: Getting started with OSC
+nav_order: 10
+has_children: false
+parent: .NET clients
+---
+
+# Getting started with OSC
```
[the remark this answers is not in the document]
Comment (kolchfa-aws) [reply _clients/OSC-dot-net.md:9]: I was going with the package name which is OSC... Let me know if I should change to OpenSearch.Client.

[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply _clients/OSC-dot-net.md:9]: Hi @kolchfa-aws,
OCS was renamed to OpenSearch.Client to avoid naming conflict in nuget.

[the remark this answers is not in the document]
Comment (kolchfa-aws) [reply _clients/OSC-dot-net.md:9]: Thank you! I will rename all instances of OSC to OpenSearch.Client.

File: _clients/dot-net.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+[OSC]({{site.url}}{{site.baseurl}}/clients/OSC-dot-net/) is a high-level .NET client on top of OpenSearch.Net. It provides strongly typed requests and responses, as well as Query DSL. It frees you from constructing raw JSON requests and parsing raw JSON responses by supplying models that parse and serialize/deserialize requests and responses automatically. OSC also exposes the OpenSearch.Net low-level client if you need it. OSC includes the following advanced functionality:
+
+- Automapping: given a C# type, OSC can infer the correct mapping to send to OpenSearch.
+- Operator overloading in queries.
+- Type and index inference.
+
+You can use both .NET clients in a console program, .NET core, ASP.NET core, or worker services.
+
+To get started with OSC, follow the [getting started guide]({{site.url}}{{site.baseurl}}/clients/OSC-dot-net#installing-the-osc-client) or a slightly [more advanced example]({{site.url}}{{site.baseurl}}/clients/OSC-dot-net/osc-example).
```
Comment (Yury-Fridlyand) [inline _clients/dot-net.md:23]: The second link is broken

Comment (kolchfa-aws) [reply _clients/dot-net.md]: Changed the name to OpenSearch.Client and fixed all links.

**Comment by kolchfa-aws** [comment]
Tracking the nuget package release in this issue: https://github.com/opensearch-project/opensearch-build/issues/2051

File: _clients/OSC-example.md [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+                            .Keyword(k => k.Name(f => f.FirstName))
+                            .Keyword(k => k.Name(f => f.LastName))))
+                            .Settings(s => s.NumberOfShards(1).NumberOfReplicas(2)));
+```
+
+## Indexing multiple documents using the bulk API
+
+In addition to indexing one document using `Index` and `IndexDocument` and indexing multiple documents using `IndexMany`, you can gain more control for indexing your documents by using `Bulk` or `BulkAll`. Indexing documents individually is inefficient because it creates an HTTP request for every document sent. The BulkAll helper frees you from handling retry, chunking or back off request functionality. It automatically retries if the request fails, backs off if the server is down, and controls how many documents are sent in one HTTP request. 
+
+In the following example, BulkAll is configured with the index name, number of back off retires, and back off time. Additionally, maximum degrees of parallelism controls the number of parallel HTTP requests containing the data. Finally, the size parameter signals how many documents are sent in one HTTP request.
```
Comment (MaxKsyunz) [inline _clients/OSC-example.md:102]:
```suggestion
In the following example, `BulkAll` is configured with the index name, number of back off retires, and back off time. Additionally, maximum degrees of parallelism controls the number of parallel HTTP requests containing the data. Finally, the size parameter signals how many documents are sent in one HTTP request. 
```

**Review by MaxKsyunz — APPROVED** [review]

**Comment by kolchfa-aws** [comment]
Thank you, @Yury-Fridlyand and @MaxKsyunz for reviewing the PR!

File: _clients/OSC-dot-net.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+# Getting started with OpenSearch.Client
+
+OpenSearch.Client is a high-level .NET client on top of OpenSearch.Net. It provides strongly typed requests and responses, as well as Query DSL. It frees you from constructing raw JSON requests and parsing raw JSON responses by providing models that parse and serialize/deserialize requests and responses automatically. OpenSearch.Client also exposes the OpenSearch.Net low-level client if you need it.
+
+This getting started guide illustrates how to connect to OpenSearch, index documents, and run queries.
+
+## Installing OpenSearch.Client
+
+To install OpenSearch.Client, download the OpenSearch.Client NuGet package and add it to your project in an IDE of your choice. In Microsoft Visual Studio:
```
Comment (lizsnyder) [inline _clients/OSC-dot-net.md:17]: Should we link off to the location? Is this it? https://www.nuget.org/packages/Terradue.OpenSearch.Client/

Comment (kolchfa-aws) [reply _clients/OSC-dot-net.md:17]: I will update the location once the package is uploaded to nuget. Thanks!

Comment (lizsnyder) [reply _clients/OSC-dot-net.md:17]: Disregard, discussed offline that it's not on Nuget yet https://github.com/opensearch-project/opensearch-build/issues/2051

File: _clients/OSC-dot-net.md [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+
+`ConnectionConfiguration` is used to pass configuration options to the low-level OpenSearch.Net client. `ConnectionSettings` inherits from `ConnectionConfiguration` and provides additional configuration options.
+To set the address of the node and the default index name for requests which don't specify the index name, create a `ConnectionSettings` object:
+
+```cs
+var node = new Uri("http://myserver:9200");
+var config = new ConnectionSettings(node).DefaultIndex("students");
+var client = new OpenSearchClient(config);
+```
+
```
Comment (cwillum) [inline _clients/OSC-dot-net.md:89]: extra space? or intentional?

File: _clients/OSC-dot-net.md [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+    {
+        Console.WriteLine($"{s.Id} {s.LastName} {s.FirstName} {s.Gpa} {s.GradYear}");
+    }
+}
+```
+
+The response contains one document, which corresponds to the correct student:
+
+`100 Santos Paulo 3.93 2021`
+## Using OpenSearch.Client methods asynchronously
```
Comment (cwillum) [inline _clients/OSC-dot-net.md:182]: space here?

File: _clients/OSC-example.md [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+layout: default
+title: More advanced features of OpenSearch.Client
+nav_order: 12
+has_children: false
+parent: .NET clients
+---
+
+# More advanced features of OpenSearch.Client
+
+The following example illustrates more advanced features of OpenSearch.Client. For a simple example, see the [getting started guide]({{site.url}}{{site.baseurl}}/clients/OSC-dot-net/). This example uses the following Student class.
```
Comment (cwillum) [inline _clients/OSC-example.md:11]: sentence cap here? "getting"

File: _clients/OSC-example.md [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+                            .Keyword(k => k.Name(f => f.FirstName))
+                            .Keyword(k => k.Name(f => f.LastName))))
+                            .Settings(s => s.NumberOfShards(1).NumberOfReplicas(2)));
+```
+
+## Indexing multiple documents using the bulk API
+
+In addition to indexing one document using `Index` and `IndexDocument` and indexing multiple documents using `IndexMany`, you can gain more control for indexing your documents by using `Bulk` or `BulkAll`. Indexing documents individually is inefficient because it creates an HTTP request for every document sent. The BulkAll helper frees you from handling retry, chunking or back off request functionality. It automatically retries if the request fails, backs off if the server is down, and controls how many documents are sent in one HTTP request. 
+
+In the following example, `BulkAll` is configured with the index name, number of back off retires, and back off time. Additionally, maximum degrees of parallelism controls the number of parallel HTTP requests containing the data. Finally, the size parameter signals how many documents are sent in one HTTP request.
```
Comment (cwillum) [inline _clients/OSC-example.md:102]: back-off "retires" or retries?

File: _clients/OSC-example.md [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+                        .From(0)
+                        .Size(10)
+                        .Scroll("1m")
+                        .Query(q => q
+                        .Bool(b => b
+                        .Filter(f => f
+                        .Term(t => t.Field(fld => fld.GradYear).Value(2022)))))
+                        .Sort(srt => srt.Ascending(f => f.LastName)));
+```
+
```
Comment (cwillum) [inline _clients/OSC-example.md:134]: extra space here?

File: _clients/OpenSearch-dot-net.md [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+layout: default
+title: OpenSearch.Net client
+nav_order: 30
+has_children: false
+parent: .NET clients
+---
+
+# OpenSearch.Net client
+
+OpenSearch.Net is a low-level .NET client that provides the foundational layer of communicating with OpenSearch. It is dependency-free, and it can handle round-robin load balancing, transport, and the basic request/response cycle. OpenSearch.Net contains all OpenSearch API endpoints as methods. When using OpenSearch.Net, you need to construct the queries yourself.
```
Comment (cwillum) [inline _clients/OpenSearch-dot-net.md:11]: "of communication with"?

File: _clients/dot-net.md [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+nav_order: 75
+has_children: true
+has_toc: false
+---
+
+# .NET clients
+
+OpenSearch has two .NET clients: a low-level [OpenSearch.Net]({{site.url}}{{site.baseurl}}/clients/OpenSearch-dot-net/) client, and a high-level [OpenSearch.Client]({{site.url}}{{site.baseurl}}/clients/OSC-dot-net/) client.
+
+[OpenSearch.Net]({{site.url}}{{site.baseurl}}/clients/OpenSearch-dot-net/) is a low-level .NET client that provides the foundational layer of communicating with OpenSearch. It is dependency-free, and it can handle round-robin load balancing, transport, and the basic request/response cycle. OpenSearch.Net contains methods for all OpenSearch API endpoints.
```
Comment (cwillum) [inline _clients/dot-net.md:13]: "... of communication with OpenSearch"? Maybe "...of communicating with..." is conventional in this space. But caught my eye.

**Review by cwillum — APPROVED** [review]
Looks good. Just a few small comments.

File: _clients/OSC-dot-net.md (hunk 5 above)
Comment (kolchfa-aws) [reply _clients/OSC-dot-net.md:89]: Removed

File: _clients/OSC-dot-net.md (hunk 6 above)
Comment (kolchfa-aws) [reply _clients/OSC-dot-net.md:182]: Added

File: _clients/OSC-example.md (hunk 7 above)
Comment (kolchfa-aws) [reply _clients/OSC-example.md:11]: Changed

File: _clients/OSC-example.md (hunk 8 above)
Comment (kolchfa-aws) [reply _clients/OSC-example.md:102]: Good catch - thank you!

File: _clients/OSC-example.md (hunk 9 above)
Comment (kolchfa-aws) [reply _clients/OSC-example.md:134]: Removed

File: _clients/dot-net-conventions.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+When first created, a node is master eligible and its `HoldsData` property is set to true. 
+The `AbsolutePath` property of the node created above is `"/opensearch/"`: a trailing forward slash is appended so that the paths can be easily combined. If not specified, the default `Port` is 80.
+
+Nodes are considered equal if they have the same endpoint. Metadata is not taken into account when checking nodes for equality.
+{: .note}
+
+## Connection pools
+
+Connecttion pools are instances of `IConnectionPool` and are responsible for managing the nodes in the OpenSearch cluster. We recommend creating a [singleton client](#registering-the-osc-client-as-a-singleton) with a single `ConnectionSettings` object. The lifetime of both the client and its `ConnectionSettings` is the lifetime of the application.
```
Comment (hdhalter) [inline _clients/dot-net-conventions.md:40]: Connection pools (extra t in first instance)

File: _clients/dot-net-conventions.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+The following are connection pool types.
+
+- **SingleNodeConnectionPool**
+
+`SingleNodeConnectionPool` is the default connection pool that is used if no connection pool is passed to the `ConnectionSettings` constructor. Use a `SingleNodeConnectionPool` if you have only one node in the cluster or if your cluster has a load balancer as an entry point. `SingleNodeConnectionPool` does not support sniffing or pinging and does not mark nodes dead or alive.
```
Comment (hdhalter) [inline _clients/dot-net-conventions.md:46]:
```suggestion
`SingleNodeConnectionPool` is the default connection pool that is used if no connection pool is passed to the `ConnectionSettings` constructor. Use `SingleNodeConnectionPool` if you have only one node in the cluster or if your cluster has a load balancer as an entry point. `SingleNodeConnectionPool` does not support sniffing or pinging and does not mark nodes dead or alive. 
```

File: _clients/OpenSearch-dot-net.md (hunk 10 above)
Comment (JeffH-AWS) [reply _clients/OpenSearch-dot-net.md:11]: I've sat here and read the sentence 10 times both ways.  Is it wrong the way Fanit has it written?  It sounds fine to me and I tried googling a bit but didn't see anything definitive either way (although I didn't look super long, to be fair).

**Review by JeffH-AWS — APPROVED** [review]
Looks good great work Fanit!

**Review by hdhalter — COMMENTED** [review]
Just a few minor suggestions. Nice job - looks great!

File: _clients/OSC-dot-net.md [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+layout: default
+title: Getting started with OpenSearch.Client
+nav_order: 10
+has_children: false
+parent: .NET clients
+---
+
+# Getting started with OpenSearch.Client
+
+OpenSearch.Client is a high-level .NET client on top of OpenSearch.Net. It provides strongly typed requests and responses, as well as Query DSL. It frees you from constructing raw JSON requests and parsing raw JSON responses by providing models that parse and serialize/deserialize requests and responses automatically. OpenSearch.Client also exposes the OpenSearch.Net low-level client if you need it.
```
[the remark this answers is not in the document]
Comment (hdhalter) [reply _clients/OSC-dot-net.md:11]: Never mind... I'm not reading it in the right order. I understand that .dot-net.md file is first, so it makes more sense now!

File: _clients/dot-net-conventions.md (hunk 13 above)
Comment (kolchfa-aws) [reply _clients/dot-net-conventions.md:46]: @hdhalter - thank you for the feedback. I implemented your comments.

File: _clients/dot-net-conventions.md [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+If a request does not succeed, it is automatically retried. By default, the number of retries is the number of nodes in your cluster known to OpenSearch.Client. The number of retries is also limited by the timeout parameter, so OpenSearch.Client retries requests as many times as possible within the timeout period. 
+
+To set the maximum number of retries, specify this number in the `MaximumRetries` property on the `ConnectionSettings` object.
+
+```cs
+var settings = new ConnectionSettings(connectionPool).MaximumRetries(5);
+```
+
+You can also set a `RequestTimeout` that specifies a timeout for one single request, and a `MaxRetryTimeout` that specifies the time limit for all retry attempts. In the example below, `RequestTimeout` is set to 4 seconds, and `MaxRetryTimeout` is set to 12 seconds, so the maximum number of attempts for a query is 3.
```
Comment (natebower) [inline _clients/dot-net-conventions.md:78]: Replace "one" with "a". Remove the comma after "request".

File: _clients/dot-net-conventions.md [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+```cs
+var settings = new ConnectionSettings(connectionPool)
+            .RequestTimeout(TimeSpan.FromSeconds(4))
+            .MaxRetryTimeout(TimeSpan.FromSeconds(12));
+```
+
+## Failover
+
+If you are using a connection pool with multiple nodes, a request is retried if it returns a 502 (Bad Gateway), 503 (Service Unavailable) or 504 (Gateway Timeout) HTTP error response code. If the response code is an error code in the 400–501 or 505–599 range, the request is not retried.
```
Comment (natebower) [inline _clients/dot-net-conventions.md:88]: Insert the serial comma after "(Service Unavailable)". If we are referring to two ranges, "range" should be "ranges".

File: _clients/dot-net-conventions.md [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+
+A response is considered valid if the response code is in the 2xx range, or the response code has one of the expected values for this request. For example, 404 (Not Found) is a valid response for a request that checks if an index exists.
```
Comment (natebower) [inline _clients/dot-net-conventions.md:90]: Remove the first comma. Replace the last instance of "if" with "whether".

File: _clients/dot-net.md [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+layout: default
+title: .NET clients
+nav_order: 75
+has_children: true
+has_toc: false
+---
+
+# .NET clients
+
+OpenSearch has two .NET clients: a low-level [OpenSearch.Net]({{site.url}}{{site.baseurl}}/clients/OpenSearch-dot-net/) client, and a high-level [OpenSearch.Client]({{site.url}}{{site.baseurl}}/clients/OSC-dot-net/) client.
```
Comment (natebower) [inline _clients/dot-net.md:11]: Remove the comma.

File: _clients/dot-net.md [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+
+[OpenSearch.Net]({{site.url}}{{site.baseurl}}/clients/OpenSearch-dot-net/) is a low-level .NET client that provides the foundational layer of communicating with OpenSearch. It is dependency-free, and it can handle round-robin load balancing, transport, and the basic request/response cycle. OpenSearch.Net contains methods for all OpenSearch API endpoints.
+
+[OpenSearch.Client]({{site.url}}{{site.baseurl}}/clients/OSC-dot-net/) is a high-level .NET client on top of OpenSearch.Net. It provides strongly typed requests and responses, as well as Query DSL. It frees you from constructing raw JSON requests and parsing raw JSON responses by supplying models that parse and serialize/deserialize requests and responses automatically. OpenSearch.Client also exposes the OpenSearch.Net low-level client if you need it. OpenSearch.Client includes the following advanced functionality:
```
Comment (natebower) [inline _clients/dot-net.md:15]: Remove the comma after "responses".

File: _clients/dot-net.md [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+
+- Automapping: given a C# type, OpenSearch.Client can infer the correct mapping to send to OpenSearch.
```
Comment (natebower) [inline _clients/dot-net.md:17]: Capitalize "Given".

File: _clients/dot-net.md [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+- Operator overloading in queries.
+- Type and index inference.
+
+You can use both .NET clients in a console program, .NET core, ASP.NET core, or worker services.
```
Comment (natebower) [inline _clients/dot-net.md:21]: Rephrase this as "...in a console program, a .NET core, an ASP.NET core, or in worker services."

**Review by natebower — COMMENTED** [review]
@kolchfa-aws Please see my comments and let me know if you have any questions. Thanks!

File: _clients/OSC-dot-net.md [hunk 22]
```diff
@@ -0,0 +1,329 @@
...
+namespace NetClientProgram;
+
+internal class Program
+{
+    private static IOpenSearchClient osClient = new OpenSearchClient();
+
+    public static void Main(string[] args)
+    {       
+        Console.WriteLine("Indexing one student......");
+        var student = new Student { Id = 100,
```
Comment (lizsnyder) [inline _clients/OSC-dot-net.md:243]: When I copy this program into my IDE I get an error "The type or namespace name 'Student' could not be found (are you missing a using directive or an assembly reference?)"

Do we need to define student somewhere? ex. 
```
public class Student
{
   ....
}
```

Comment (kolchfa-aws) [reply _clients/OSC-dot-net.md:243]: Yes, you should copy the class Student (defined at the beginning of the file) to the same project - in a separate file from your program
